### PR TITLE
Avoid react error when changing locale

### PR DESCRIPTION
**Pull request by matmilbury**
I use `remix-i18next` with `i18next-locize-backend`.

Everything works properly except when changing the locale on the client for the first time I get this error:
CleanShot 2021-12-16 at 14 58 35@2x
Subsequent changes don't cause that error anymore.

I don't have a deep enough understanding to explain the error, but [I've found](https://flaviocopes.com/react-update-while-rendering-different-component/) that using `useEffect` in combination with `useMemo` fixes the issue.

### CODE REVIEW

**Comment by danestves** [comment]
+1 this is happening to me too

**Comment by hauptrolle** [comment]
@sergiodxa Any objections to merge this one? I am getting this error too.

**Comment by Zloka** [comment]
What is the purpose behind the `useMemo` call on line 45 (and the old `useMemo` call, I guess)?

File: src/react.tsx [hunk 1]
```diff
@@ -28,12 +35,20 @@ export function useRemixI18Next(locale: string) {
       )
   );
 
-  useMemo(() => {
+  const handleLocaleUpdate = useCallback(() => {
```
Comment (matmilbury) [inline src/react.tsx:38]: preserves `handleLocaleUpdate` function identity across renders unless dependencies change.

File: src/react.tsx [hunk 2]
```diff
@@ -28,12 +35,20 @@ export function useRemixI18Next(locale: string) {
...
     i18next.changeLanguage(locale);
     for (let [namespace, messages] of Object.entries(namespaces)) {
       i18next.addResourceBundle(locale, namespace, messages);
     }
   }, [i18next, namespaces, locale]);
+
+  useMemo(() => {
+    handleLocaleUpdate();
+  }, []);
```
Comment (matmilbury) [inline src/react.tsx:47]: calls `handleLocaleUpdate` on the first render

File: src/react.tsx [hunk 3]
```diff
@@ -28,12 +35,20 @@ export function useRemixI18Next(locale: string) {
...
+
+  useEffect(() => {
+    handleLocaleUpdate();
+  }, [handleLocaleUpdate]);
```
Comment (matmilbury) [inline src/react.tsx:51]: calls `handleLocaleUpdate` on subsequent renders but only when the identity of `handleLocaleUpdate` changes.

**Comment by matmilbury** [comment]
@Zloka I've added inline comments with explanations.

**Comment by Zloka** [comment]
> @Zloka I've added inline comments with explanations.

Thanks! Doesn't the `useEffect` call it on the initial render too, though?

**Comment by matmilbury** [comment]
Not really - it's different on the server and in the browser.

SSR: `useMemo` calls `handleLocaleUpdate` during the first rendering. `useEffect` is never triggered.

Browser: `useMemo` calls `handleLocaleUpdate` _during_ the first rendering. Then, the DOM is rendered. Immediately after the DOM is rendered, `useEffect` calls `handleLocaleUpdate` again.

See https://stackblitz.com/edit/react-ts-2vppyt?file=index.tsx

**Comment by Zloka** [comment]
Ah, I see, still a bit new so I failed to consider the SSR-case 🙂 

Do you happen to know if calling `handleLocaleUpdate` _during_ the first render is beneficial somehow?

**Comment by matmilbury** [comment]
@Zloka I'm happy to help 🙂 

Calling `handleLocaleUpdate` during the first render is necessary for SSR to render correct language.

**Comment by Zloka** [comment]
> @Zloka I'm happy to help 🙂
> 
> Calling `handleLocaleUpdate` during the first render is necessary for SSR to render correct language.

Of course 🤦  Thanks again!

File: src/react.tsx (hunk 1 above)
Comment (sergiodxa) [inline src/react.tsx:38]:
```suggestion
  let handleLocaleUpdate = useCallback(() => {
```

**Review by sergiodxa — APPROVED** [review]

**Pull request merged**